lib/mongo/collection/view/writable.rb

```diff
@@ -233,12 +236,23 @@ def replace_one(replacement, opts = {})
           update_doc = { Operation::Q => filter,
                          Operation::U => replacement,
                         }
-          if opts[:upsert]
-            update_doc['upsert'] = true
-          end
+          update_doc['upsert'] = true if opts[:upsert]
+          update_doc['hint'] = opts[:hint] if opts[:hint]
+
           with_session(opts) do |session|
             write_concern = write_concern_with_session(session)
             write_with_retry(session, write_concern) do |server, txn_num|
+
+              if server.max_wire_version < 5 && update_doc['hint']
+                raise Error::UnsupportedCommandOption.new(
+                  option_name: 'hint',
+                  command_name: 'replace_one',
+                  required_server_version: '3.4',
+                  required_wire_version: '5',
+                  server_max_wire_version: server.max_wire_version
+                )
+              end
```

Comment:
This code is not very DRY, but the operations layer isn't very DRY in general. Would it be preferable for me to try and abstract this logic into a helper module?